src/Projects/Modules/HorribleSubs/Server/Module.HorribleSubs.Cida/Services/DownloadService.cs

```diff
@@ -0,0 +1,150 @@
+        private readonly ConcurrentDictionary<string, CreateDownloaderContext> requestedDownloads;
+        private readonly ConcurrentDictionary<string, DownloadProgress> downloadStatus;
+        private readonly HorribleSubsDbContext context;
+        private readonly IFtpClient ftpClient;
+
+        public IReadOnlyDictionary<string, DownloadProgress> CurrentDownloadStatus => this.downloadStatus.ToDictionary(pair => pair.Key, pair => pair.Value);
+
+        public DownloadService(string host, int port, HorribleSubsDbContext context, IFtpClient ftpClient)
+        {
+            string name = "ad_" + Guid.NewGuid();
+            this.requestedDownloads = new ConcurrentDictionary<string, CreateDownloaderContext>();
+            this.downloadStatus = new ConcurrentDictionary<string, DownloadProgress>();
+            this.ircClient = new IrcClient.Clients.IrcClient(host, port, name, name, name, this.tempFolder);
+
+
+            this.ircClient.DownloadRequested += downloader =>
+            {
+                if (this.requestedDownloads.TryGetValue(downloader.Filename, out var context))
+                {
+                    context.Downloader = downloader;
+                    context.ManualResetEvent.Set();
+                }
+            };
+
+            this.context = context;
+            this.ftpClient = ftpClient;
+        }
+
+        public async Task CreateDownloader(DownloadRequest downloadRequest)
+        {
+            if ((await this.context.Downloads.FindAsync(downloadRequest.FileName)) != null)
+            {
+                return;
+            }
+
+            if (!this.ircClient.IsConnected)
+            {
+                this.ircClient.Connect();
+            }
+
+            var createDownloaderContext = new CreateDownloaderContext()
+            {
+                Filename = downloadRequest.FileName,
+            };
+            var dccDownloaderTask = new Task<DccDownloader>(() =>
+            {
+                createDownloaderContext.ManualResetEvent.Wait();
+                return createDownloaderContext.Downloader;
+
+            }, TaskCreationOptions.LongRunning);
+
+            if (this.requestedDownloads.TryAdd(downloadRequest.FileName, createDownloaderContext))
+            {
+                this.ircClient.SendMessage($"xdcc send #{downloadRequest.PackageNumber}", downloadRequest.BotName);
+            }
+            dccDownloaderTask.Start();
+            var downloader = await dccDownloaderTask;
+
+            await this.context.Downloads.AddAsync(new Download()
+            {
+                Name = downloader.Filename,
+                Size = downloader.Filesize,
+                DownloadStatus = DownloadStatus.Downloading,
+            });
+
+            await this.context.SaveChangesAsync();
+
+#pragma warning disable CS4014 // Because this call is not awaited, execution of the current method continues before the call is completed
+            Task.Run(async () => await this.DownloadFile(downloader));
```

Comment:
I feel like this one is connected with my comment in [DccDownloader.cs](https://github.com/Denrage/Cida/pull/19/files/e5fc746fb98b8b06656e95b2914f254b0759de3c#diff-1717a216e93237ff1b316b9092d6633b)

Reply:
The goal of the gRPC endpoint who is calling this is to only start the download. It is intended that it won't block till the download is done bc that can take a long time and it's not needed to have a continuous connection with the client till the download is done